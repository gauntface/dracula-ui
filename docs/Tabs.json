{
  "description": "Tabs are horizontal navigation elements used to display\ncontent within the same page.",
  "displayName": "Tabs",
  "methods": [],
  "props": {
    "color": {
      "defaultValue": null,
      "description": "The Dracula UI color for the Tabs.",
      "name": "color",
      "parent": {
        "fileName": "src/components/Tabs/Tabs.tsx",
        "name": "TabsProps"
      },
      "required": true,
      "type": {
        "name": "\"black\" | \"white\" | \"cyan\" | \"green\" | \"orange\" | \"pink\" | \"purple\" | \"red\" | \"yellow\" | \"purpleCyan\" | \"yellowPink\" | \"cyanGreen\" | \"pinkPurple\" | \"animated\""
      }
    }
  }
}
