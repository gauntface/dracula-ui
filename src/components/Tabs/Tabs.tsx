import cx from 'classnames/dedupe'
import React, { HTMLAttributes } from 'react'
import { ColorMap } from '../../base/colors'
import {
  marginClasses,
  MarginPropType,
  spacingClasses,
  SpacingPropType
} from '../../base/spacing'

export const tabsColors: Partial<ColorMap> = {
  white: 'drac-tabs-white',
  cyan: 'drac-tabs-cyan',
  green: 'drac-tabs-green',
  orange: 'drac-tabs-orange',
  pink: 'drac-tabs-pink',
  purple: 'drac-tabs-purple',
  red: 'drac-tabs-red',
  yellow: 'drac-tabs-yellow'
}

/**
 * Tabs Props
 */
export interface TabsProps extends HTMLAttributes<HTMLUListElement> {
  /** The Dracula UI color for the Tabs. */
  color: keyof typeof tabsColors

  /** Dracula UI standard spacing properties. */
  spacing?: SpacingPropType

  /** Dracula UI standard margin properties. */
  margin?: MarginPropType
}

/**
 * Tabs are horizontal navigation elements used to display
 * content within the same page.
 */
export const Tabs: React.FC<TabsProps> = (props: TabsProps) => {
  const finalProps = {
    ...props,
    className: cx(
      'drac-tabs',
      props.color && tabsColors[props.color],
      ...spacingClasses(props.spacing),
      ...marginClasses(props.margin)
    )
  }

  return <ul {...finalProps} />
}

Tabs.displayName = 'Tabs'
