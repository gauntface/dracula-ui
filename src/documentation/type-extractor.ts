import * as ts from 'typescript'

export function extractWithComment(
  fileNames: string[],
  options: ts.CompilerOptions
): void {
  const program = ts.createProgram(fileNames, options)
  const checker: ts.TypeChecker = program.getTypeChecker()

  for (const sourceFile of program.getSourceFiles()) {
    if (!sourceFile.isDeclarationFile) {
      ts.forEachChild(sourceFile, visit)
    }
  }

  function visit(node: ts.Node) {
    const count = node.getChildCount()

    if (count > 0) {
      ts.forEachChild(node, visit)
    }

    if (ts.isInterfaceDeclaration(node) && node.name) {
      const symbol = checker.getSymbolAtLocation(node.name)
      if (symbol) {
        return serializeSymbol(symbol)
      }
    }
  }

  function serializeSymbol(symbol: ts.Symbol) {
    const s = {
      name: symbol.getName(),
      comment: ts.displayPartsToString(symbol.getDocumentationComment(checker)),
      type: checker.typeToString(
        checker.getTypeOfSymbolAtLocation(symbol, symbol.valueDeclaration!)
      )
    }
    console.log(s)
    return s
  }
}

console.log(extractWithComment(['src/components/Avatar/Avatar.tsx'], {}))
