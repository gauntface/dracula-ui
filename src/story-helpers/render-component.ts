import testRenderer from 'react-test-renderer'

export type ComponentExample = {
  code: ReturnType<testRenderer.ReactTestRenderer['toJSON']>
  docs: string
  title: string
}

export type SnapshotBuilder = {
  comp: () => React.ReactElement
  docs: string
  title: string
}

export function buildSnapshot(
  title: string,
  comp: () => React.ReactElement,
  docs: string = ''
): SnapshotBuilder {
  return {
    title,
    comp,
    docs: docs.trim()
  }
}

export function renderSnapshot(
  title: string,
  comp: React.ReactElement,
  docs: string = ''
): ComponentExample {
  const rendered = testRenderer.create(comp)
  const code = rendered.toJSON()
  return {
    title,
    code,
    docs
  }
}
