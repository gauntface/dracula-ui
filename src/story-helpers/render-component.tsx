import React from 'react'
import Enzyme, { mount,  shallow } from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'
import prettier from 'prettier'
import toJson from "enzyme-to-json";

Enzyme.configure({ adapter: new Adapter() })

export type ComponentExample = {
  html: string
  docs: string
  title: string
  react?: any
}

export type SnapshotBuilder = {
  comp: () => React.ReactElement
  docs: string
  title: string
}

export function buildSnapshot(
  title: string,
  comp: () => React.ReactElement,
  docs: string = ''
): SnapshotBuilder {
  return {
    title,
    comp,
    docs: docs.trim()
  }
}

export function renderSnapshot(
  title: string,
  comp: React.ReactElement,
  docs: string = ''
): ComponentExample {

  
  const html = mount(comp).html()
  const react = toJson(shallow(<div>{comp}</div>))

  const prettyHTML = prettier.format(html, {
    parser: 'html'
  })

  return {
    title,
    html: prettyHTML,
    react,
    docs
  }
}
