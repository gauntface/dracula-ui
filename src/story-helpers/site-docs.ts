import {
  renderSnapshot,
  SnapshotBuilder
} from '@/story-helpers/render-component'
import React from 'react'

// import * as serializer from 'jest-snapshot-serializer-raw'

interface Documentation {
  basicUsage: () => SnapshotBuilder
  variations: () => Array<SnapshotBuilder>
}

export function siteDocs<T>(
  componentClass: React.FunctionComponent<T>,
  documentation: Documentation
) {
  describe(`Site: ${componentClass.displayName ?? componentClass.name}`, () => {
    test('Basic Usage', () => {
      const variation = documentation.basicUsage()
      const snapshot = renderSnapshot(
        variation.title,
        variation.comp(),
        variation.docs
      )
      expect(snapshot).toMatchSnapshot()
    })

    const variations = documentation.variations()

    variations.forEach(variation => {
      test(variation.title, () => {
        const snapshot = renderSnapshot(
          variation.title,
          variation.comp(),
          variation.docs
        )
        expect(snapshot).toMatchSnapshot()
      })
    })
  })
}
