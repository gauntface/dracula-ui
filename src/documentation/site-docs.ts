import {
  ComponentExample,
  renderSnapshot,
  SnapshotBuilder
} from '../story-helpers/render-component'
import React from 'react'
import fs from 'fs-extra'
import { pretty } from '../story-helpers/pretty'
import { getDocGen } from '../documentation/docgen/doc-generator'
import { toDSP } from '../documentation/dsp/component-generator'

interface Documentation {
  basicUsage: () => SnapshotBuilder
  variations: () => Array<SnapshotBuilder>
}

export function siteDocs<T>(
  componentClass: React.FunctionComponent<T>,
  documentation: Documentation
) {
  const name = componentClass.displayName ?? componentClass.name

  describe(`Site: ${name}`, () => {
    let examples: Record<string, ComponentExample> = {}

    afterAll(async () => {
      const docGen = getDocGen(name)
      const dsp = toDSP(name, examples, docGen)
      const dspPath = `${process.cwd()}/dsp/data/components/${name}.json`
      await fs.writeFile(dspPath, pretty(JSON.stringify(dsp), 'json'))
    })

    test('Basic Usage', () => {
      const variation = documentation.basicUsage()
      const snapshot = renderSnapshot(
        variation.title,
        variation.comp,
        variation.docs
      )
      expect(snapshot).toMatchSnapshot()
      examples = {
        ...examples,
        basic: snapshot
      }
    })

    const variations = documentation.variations()

    variations.forEach((variation) => {
      test(variation.title, () => {
        const snapshot = renderSnapshot(
          variation.title,
          variation.comp,
          variation.docs
        )
        expect(snapshot).toMatchSnapshot()

        examples = {
          ...examples,
          [variation.title]: snapshot
        }
      })
    })
  })
}
