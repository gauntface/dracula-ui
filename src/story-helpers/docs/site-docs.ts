import {
  ComponentExample,
  renderSnapshot,
  SnapshotBuilder
} from '@/story-helpers/render-component'
import React from 'react'
import fs from 'fs'
import { pretty } from '@/story-helpers/pretty'

interface Documentation {
  basicUsage: () => SnapshotBuilder
  variations: () => Array<SnapshotBuilder>
}

export function siteDocs<T>(
  componentClass: React.FunctionComponent<T>,
  documentation: Documentation
) {
  const name = componentClass.displayName ?? componentClass.name

  describe(`Site: ${name}`, () => {
    let examples: Record<string, ComponentExample> = {}

    afterAll(() => {
      const path = `${process.cwd()}/examples/${name}.json`
      fs.writeFileSync(path, pretty(JSON.stringify(examples), 'json'))
    })

    test('Basic Usage', () => {
      const variation = documentation.basicUsage()
      const snapshot = renderSnapshot(
        variation.title,
        variation.comp,
        variation.docs
      )
      expect(snapshot).toMatchSnapshot()
      examples = {
        ...examples,
        basic: snapshot
      }
    })

    const variations = documentation.variations()

    variations.forEach((variation) => {
      test(variation.title, () => {
        const snapshot = renderSnapshot(
          variation.title,
          variation.comp,
          variation.docs
        )
        expect(snapshot).toMatchSnapshot()

        examples = {
          ...examples,
          [variation.title]: snapshot
        }
      })
    })
  })
}
