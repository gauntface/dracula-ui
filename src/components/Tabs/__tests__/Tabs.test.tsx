import { Tabs } from '@/components/Tabs/Tabs'
import {
  TabsColors,
  TabsNormal
} from '@/components/Tabs/__stories__/Tabs.stories'
import { docs } from '@/documentation/site-docs'
import { snapshot } from '../../../documentation/render-component'

docs(Tabs, {
  basic: () => snapshot('Usage', TabsNormal),
  variations: () => snapshot('Colors', TabsColors)
})
