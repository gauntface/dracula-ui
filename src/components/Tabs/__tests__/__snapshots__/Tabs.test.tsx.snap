// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Site: Tabs Basic Usage 1`] = `
Object {
  "docs": "",
  "html": "<ul color=\\"purple\\" class=\\"drac-tabs drac-tabs-purple\\">
  <li class=\\"drac-tab drac-tab-active\\">
    <a class=\\"drac-tab-link drac-text\\" href=\\"#\\">Item One</a>
  </li>
  <li class=\\"drac-tab\\">
    <a class=\\"drac-tab-link drac-text\\" href=\\"#\\">Item Two</a>
  </li>
  <li class=\\"drac-tab\\">
    <a class=\\"drac-tab-link drac-text\\" href=\\"#\\">Item Three</a>
  </li>
</ul>
",
  "react": "<Tabs color=\\"purple\\">
  <li className=\\"drac-tab drac-tab-active\\">
    <a className=\\"drac-tab-link drac-text\\" href=\\"#\\">
      Item One
    </a>
  </li>
  <li className=\\"drac-tab\\">
    <a className=\\"drac-tab-link drac-text\\" href=\\"#\\">
      Item Two
    </a>
  </li>
  <li className=\\"drac-tab\\">
    <a className=\\"drac-tab-link drac-text\\" href=\\"#\\">
      Item Three
    </a>
  </li>
</Tabs>;
",
  "title": "Usage",
}
`;

exports[`Site: Tabs Colors 1`] = `
Object {
  "docs": "",
  "html": "<div style=\\"display: flex; flex-direction: column\\">
  <ul color=\\"white\\" class=\\"drac-tabs drac-tabs-white\\">
    <li class=\\"drac-tab drac-tab-active\\">
      <a class=\\"drac-tab-link drac-text\\" href=\\"#\\">Item One</a>
    </li>
    <li class=\\"drac-tab\\">
      <a class=\\"drac-tab-link drac-text\\" href=\\"#\\">Item Two</a>
    </li>
    <li class=\\"drac-tab\\">
      <a class=\\"drac-tab-link drac-text\\" href=\\"#\\">Item Three</a>
    </li>
  </ul>
  <ul color=\\"cyan\\" class=\\"drac-tabs drac-tabs-cyan\\">
    <li class=\\"drac-tab drac-tab-active\\">
      <a class=\\"drac-tab-link drac-text\\" href=\\"#\\">Item One</a>
    </li>
    <li class=\\"drac-tab\\">
      <a class=\\"drac-tab-link drac-text\\" href=\\"#\\">Item Two</a>
    </li>
    <li class=\\"drac-tab\\">
      <a class=\\"drac-tab-link drac-text\\" href=\\"#\\">Item Three</a>
    </li>
  </ul>
  <ul color=\\"green\\" class=\\"drac-tabs drac-tabs-green\\">
    <li class=\\"drac-tab drac-tab-active\\">
      <a class=\\"drac-tab-link drac-text\\" href=\\"#\\">Item One</a>
    </li>
    <li class=\\"drac-tab\\">
      <a class=\\"drac-tab-link drac-text\\" href=\\"#\\">Item Two</a>
    </li>
    <li class=\\"drac-tab\\">
      <a class=\\"drac-tab-link drac-text\\" href=\\"#\\">Item Three</a>
    </li>
  </ul>
  <ul color=\\"orange\\" class=\\"drac-tabs drac-tabs-orange\\">
    <li class=\\"drac-tab drac-tab-active\\">
      <a class=\\"drac-tab-link drac-text\\" href=\\"#\\">Item One</a>
    </li>
    <li class=\\"drac-tab\\">
      <a class=\\"drac-tab-link drac-text\\" href=\\"#\\">Item Two</a>
    </li>
    <li class=\\"drac-tab\\">
      <a class=\\"drac-tab-link drac-text\\" href=\\"#\\">Item Three</a>
    </li>
  </ul>
  <ul color=\\"pink\\" class=\\"drac-tabs drac-tabs-pink\\">
    <li class=\\"drac-tab drac-tab-active\\">
      <a class=\\"drac-tab-link drac-text\\" href=\\"#\\">Item One</a>
    </li>
    <li class=\\"drac-tab\\">
      <a class=\\"drac-tab-link drac-text\\" href=\\"#\\">Item Two</a>
    </li>
    <li class=\\"drac-tab\\">
      <a class=\\"drac-tab-link drac-text\\" href=\\"#\\">Item Three</a>
    </li>
  </ul>
  <ul color=\\"purple\\" class=\\"drac-tabs drac-tabs-purple\\">
    <li class=\\"drac-tab drac-tab-active\\">
      <a class=\\"drac-tab-link drac-text\\" href=\\"#\\">Item One</a>
    </li>
    <li class=\\"drac-tab\\">
      <a class=\\"drac-tab-link drac-text\\" href=\\"#\\">Item Two</a>
    </li>
    <li class=\\"drac-tab\\">
      <a class=\\"drac-tab-link drac-text\\" href=\\"#\\">Item Three</a>
    </li>
  </ul>
  <ul color=\\"red\\" class=\\"drac-tabs drac-tabs-red\\">
    <li class=\\"drac-tab drac-tab-active\\">
      <a class=\\"drac-tab-link drac-text\\" href=\\"#\\">Item One</a>
    </li>
    <li class=\\"drac-tab\\">
      <a class=\\"drac-tab-link drac-text\\" href=\\"#\\">Item Two</a>
    </li>
    <li class=\\"drac-tab\\">
      <a class=\\"drac-tab-link drac-text\\" href=\\"#\\">Item Three</a>
    </li>
  </ul>
  <ul color=\\"yellow\\" class=\\"drac-tabs drac-tabs-yellow\\">
    <li class=\\"drac-tab drac-tab-active\\">
      <a class=\\"drac-tab-link drac-text\\" href=\\"#\\">Item One</a>
    </li>
    <li class=\\"drac-tab\\">
      <a class=\\"drac-tab-link drac-text\\" href=\\"#\\">Item Two</a>
    </li>
    <li class=\\"drac-tab\\">
      <a class=\\"drac-tab-link drac-text\\" href=\\"#\\">Item Three</a>
    </li>
  </ul>
</div>
",
  "react": "<div style={{ display: \\"flex\\", flexDirection: \\"column\\" }}>
  <Tabs color=\\"white\\">
    <li className=\\"drac-tab drac-tab-active\\">
      <a className=\\"drac-tab-link drac-text\\" href=\\"#\\">
        Item One
      </a>
    </li>
    <li className=\\"drac-tab\\">
      <a className=\\"drac-tab-link drac-text\\" href=\\"#\\">
        Item Two
      </a>
    </li>
    <li className=\\"drac-tab\\">
      <a className=\\"drac-tab-link drac-text\\" href=\\"#\\">
        Item Three
      </a>
    </li>
  </Tabs>
  <Tabs color=\\"cyan\\">
    <li className=\\"drac-tab drac-tab-active\\">
      <a className=\\"drac-tab-link drac-text\\" href=\\"#\\">
        Item One
      </a>
    </li>
    <li className=\\"drac-tab\\">
      <a className=\\"drac-tab-link drac-text\\" href=\\"#\\">
        Item Two
      </a>
    </li>
    <li className=\\"drac-tab\\">
      <a className=\\"drac-tab-link drac-text\\" href=\\"#\\">
        Item Three
      </a>
    </li>
  </Tabs>
  <Tabs color=\\"green\\">
    <li className=\\"drac-tab drac-tab-active\\">
      <a className=\\"drac-tab-link drac-text\\" href=\\"#\\">
        Item One
      </a>
    </li>
    <li className=\\"drac-tab\\">
      <a className=\\"drac-tab-link drac-text\\" href=\\"#\\">
        Item Two
      </a>
    </li>
    <li className=\\"drac-tab\\">
      <a className=\\"drac-tab-link drac-text\\" href=\\"#\\">
        Item Three
      </a>
    </li>
  </Tabs>
  <Tabs color=\\"orange\\">
    <li className=\\"drac-tab drac-tab-active\\">
      <a className=\\"drac-tab-link drac-text\\" href=\\"#\\">
        Item One
      </a>
    </li>
    <li className=\\"drac-tab\\">
      <a className=\\"drac-tab-link drac-text\\" href=\\"#\\">
        Item Two
      </a>
    </li>
    <li className=\\"drac-tab\\">
      <a className=\\"drac-tab-link drac-text\\" href=\\"#\\">
        Item Three
      </a>
    </li>
  </Tabs>
  <Tabs color=\\"pink\\">
    <li className=\\"drac-tab drac-tab-active\\">
      <a className=\\"drac-tab-link drac-text\\" href=\\"#\\">
        Item One
      </a>
    </li>
    <li className=\\"drac-tab\\">
      <a className=\\"drac-tab-link drac-text\\" href=\\"#\\">
        Item Two
      </a>
    </li>
    <li className=\\"drac-tab\\">
      <a className=\\"drac-tab-link drac-text\\" href=\\"#\\">
        Item Three
      </a>
    </li>
  </Tabs>
  <Tabs color=\\"purple\\">
    <li className=\\"drac-tab drac-tab-active\\">
      <a className=\\"drac-tab-link drac-text\\" href=\\"#\\">
        Item One
      </a>
    </li>
    <li className=\\"drac-tab\\">
      <a className=\\"drac-tab-link drac-text\\" href=\\"#\\">
        Item Two
      </a>
    </li>
    <li className=\\"drac-tab\\">
      <a className=\\"drac-tab-link drac-text\\" href=\\"#\\">
        Item Three
      </a>
    </li>
  </Tabs>
  <Tabs color=\\"red\\">
    <li className=\\"drac-tab drac-tab-active\\">
      <a className=\\"drac-tab-link drac-text\\" href=\\"#\\">
        Item One
      </a>
    </li>
    <li className=\\"drac-tab\\">
      <a className=\\"drac-tab-link drac-text\\" href=\\"#\\">
        Item Two
      </a>
    </li>
    <li className=\\"drac-tab\\">
      <a className=\\"drac-tab-link drac-text\\" href=\\"#\\">
        Item Three
      </a>
    </li>
  </Tabs>
  <Tabs color=\\"yellow\\">
    <li className=\\"drac-tab drac-tab-active\\">
      <a className=\\"drac-tab-link drac-text\\" href=\\"#\\">
        Item One
      </a>
    </li>
    <li className=\\"drac-tab\\">
      <a className=\\"drac-tab-link drac-text\\" href=\\"#\\">
        Item Two
      </a>
    </li>
    <li className=\\"drac-tab\\">
      <a className=\\"drac-tab-link drac-text\\" href=\\"#\\">
        Item Three
      </a>
    </li>
  </Tabs>
</div>;
",
  "title": "Colors",
}
`;
